File: BinaryTrees/BinaryTreeRightSideView.java
package BinaryTrees;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/*Given the root of a binary tree, 
imagine yourself standing on the right side of it, 
return the values of the nodes you can see ordered from top to bottom. */

public class BinaryTreeRightSideView {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> viewableNodes = new ArrayList<>();

        if (root == null) {
            return viewableNodes;
        }

        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);

        while (!queue.isEmpty()) {
            int currentLevelLength = queue.size();
            for (int i = 0; i < currentLevelLength; i++) {
                TreeNode node = queue.remove();

                //if last node in the current level, it is viewable
                if (i == currentLevelLength - 1) {
                    viewableNodes.add(node.val);
                }

                if (node.left != null) queue.add(node.left);
                if (node.right != null) queue.add(node.right);
            }
        }

        return viewableNodes;
    }
}

File: BinaryTrees/binaryTreeRightSideView.js
/*Given the root of a binary tree, 
imagine yourself standing on the right side of it, 
return the values of the nodes you can see ordered from top to bottom. */

const rightSideView = (root) => {
    const viewableNodes = [];

    if (!root) {
        return viewableNodes;
    }

    let queue = [root];

    while (queue.length) {
        const nextQueue = [];

        //last node is queue is viewable
        viewableNodes.push(queue.at(-1).val);

        for (let i = 0; i < queue.length; i++) {
            const currentNode = queue[i];
            if (currentNode.left) nextQueue.push(currentNode.left);
            if (currentNode.right) nextQueue.push(currentNode.right);
        }

        queue = nextQueue;
    }

    return viewableNodes;
}